<launch>
  <!-- These parameters are used for the benchmark evaluation. Documentation can be found in parameter_server.cpp -->
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find rgbdslam)/rgbd_benchmark/log_eval.conf"/>
  <env name="ROSCONSOLE_FORMAT" value="[${severity}] Time:[${time}] Thread:[${thread}]: ${message}"/>
  <arg name="debug" default="false"/>
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/time"/-->
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/gdb -ex run -args"/-->
  <arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/xterm -e gdb -ex run -args"/>
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/ddd -ex run -args"/-->
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/valgrind -DELETEME-tool=callgrind -DELETEME-callgrind-out-file=/tmp/callgrind.out"/-->
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/valgrind -DELETEME-leak-check=full "/-->
  <arg unless="$(arg debug)" name="launch_prefix" value=""/>

  <!-- These parameters are defined in the calling script (run_tests.sh)-->
  <arg name="bagfile_name" />
  <arg name="feature_type" default="SIFTGPU"/>
  <arg name="ransac_iterations" default="100"/>
  <arg name="optimizer_skip_step" default="10"/>
  <arg name="observability_threshold" default="0"/>
  <arg name="use_root_sift" default="true"/>
  <arg name="adjuster_max_iterations" default="20"/>
  <arg name="detector_grid_resolution" default="2"/>
  <arg name="max_keypoints" default="900"/>
  <arg name="min_keypoints" default="0"/><!--not relevant anymore-->
  <arg name="match_candidates" default="4"/>
  <arg name="sampled_candidates" default="4"/>
  <arg name="nn_ratio" default="0.9"/>
  <arg name="gui" default="false"/>

  <node pkg="rgbdslam" type="rgbdslam" name="rgbdslam" required="true" output="screen"  launch-prefix="$(arg launch_prefix)"> 
    <!-- Input data settings-->
    <param name="config/topic_image_mono"              value="/camera/rgb/image_color"/> 
  	<param name="config/topic_image_depth"             value="/camera/depth/image"/>
  	<param name="config/topic_points"                  value=""/>
    <param name="config/camera_info_topic"             value="/camera/rgb/camera_info"/>
    <param name="config/wide_topic"                    value=""/>
    <param name="config/wide_cloud_topic"              value=""/>

    <!--Values for benchmark dataset fr1 -->
    <!--
    <param name="config/depth_camera_fx"                     value="591.1"/>
    <param name="config/depth_camera_fy"                     value="590.1"/>
    <param name="config/depth_camera_cx"                     value="331.0"/>
    <param name="config/depth_camera_cy"                     value="234.0"/>
    -->
    <!--testing the values for fr1 from http://vision.in.tum.de/data/datasets/rgbd-dataset/file_formats 
    <param name="config/depth_camera_fx"               value="517.3"/>
    <param name="config/depth_camera_fy"               value="516.5"/>
    <param name="config/depth_camera_cx"               value="318.6"/>
    <param name="config/depth_camera_cy"               value="255.3"/>
    -->
    <!-- FR2 from above url
    -->
    <param name="config/depth_camera_fx"               value="520.9"/>
    <param name="config/depth_camera_fy"               value="521.0"/>
    <param name="config/depth_camera_cx"               value="325.1"/>
    <param name="config/depth_camera_cy"               value="249.7"/>

    <!-- Odometry configuration for FR2 Pioneer SLAM
    <param name="config/odometry_tpc"                  value="/pose"/>
    <param name="config/use_robot_odom"                value="true"/>
    <param name="config/odom_frame_name"               value="odom"/>
    <param name="config/calibrate_odometry_online"     value="true"/>
    <param name="config/base_frame_name"               value="base_link"/>
    <param name="config/odometry_information_factor"   value="1"/>
    -->

    <!-- TF information settings -->
  	<param name="config/fixed_frame_name"              value="/map"/>
    <param name="config/ground_truth_frame_name"       value=""/><!--empty string if no ground truth-->
    <param name="config/fixed_camera"                  value="true"/> <!--is the kinect fixed with respect to base, or can it be moved (false makes sense only if transform betw. base_frame and openni_camera is sent via tf)-->
  	<param name="config/start_paused"                  value="false"/>
    <param name="config/store_pointclouds"             value="true"/> <!-- if, e.g., only trajectory is required, setting this to false saves lots of memory -->

    <param name="config/feature_detector_type"         value="$(arg feature_type)"/>
    <param name="config/feature_extractor_type"        value="$(arg feature_type)"/>

    <param name="config/matcher_type"                  value="FLANN"/> <!-- FLANN (not avail for ORB features), SIFTGPU (only for SIFTGPU detector)  or BRUTEFORCE-->
  	<param name="config/bagfile_name"                  value="$(arg bagfile_name)"/>
    <param name="config/nn_distance_ratio"             value="$(arg nn_ratio)"/> <!-- Feature correspondence is valid if distance to nearest neighbour is smaller than this parameter times the distance to the 2nd neighbour -->
  	<param name="config/max_keypoints"                 value="$(arg max_keypoints)"/>
  	<param name="config/min_keypoints"                 value="$(arg min_keypoints)"/>
  	<param name="config/adjuster_max_iterations"       value="$(arg adjuster_max_iterations)"/>
  	<param name="config/detector_grid_resolution"      value="$(arg detector_grid_resolution)"/>
    <param name="config/sufficient_matches"            value="10000"/><!-- Instead of matching all new descriptors against all of a previous node in one step, sufficient_matches+100 of the new descriptors are iteratively compared to all of the previous node until sufficient_matches are found. Setting this parameter low (e.g. 2x min_matches) speeds up comparisons of frames with many matches, but with a potential loss of accuracy, as the transformation is estimated from less features. Set it to max_keypoints for maximum accuracy -->
    <param name="config/min_translation_meter"         value="-1.0"/><!--disabled -->
    <param name="config/min_rotation_degree"           value="-1.0"/><!--disabled -->
    <param name="config/max_translation_meter"         value="100.0"/><!-- transformations with motion more than this per second!, will be omitted -->
    <param name="config/max_rotation_degree"           value="900"/><!-- transformations with motion more than this per second!, will be omitted -->
  	<param name="config/min_time_reported"             value="0.005"/>

    <param name="config/predecessor_candidates"        value="$(arg match_candidates)"/><!--search through this many immediate predecessor nodes for corrspondences -->
    <param name="config/neighbor_candidates"           value="$(arg match_candidates)"/><!--search through this many graph neighbour nodes for corrspondences -->
    <param name="config/min_sampled_candidates"        value="$(arg sampled_candidates)"/><!--search through this many uniformly sampled nodes for corrspondences -->
    <param name="config/max_connections"               value="-1"/><!-- stop after this many succesfully found spation relations -->
  	
  	<param name="config/drop_async_frames"             value="false"/>
  	<param name="config/max_matches"                   value="300"/>
  	<param name="config/max_dist_for_inliers"          value="2.0"/>
  	<param name="config/ransac_iterations"             value="$(arg ransac_iterations)"/><!-- these are fast, so high values are ok -->
  	<param name="config/use_gui"                       value="$(arg gui)"/>
  	<param name="config/use_glwidget"                  value="$(arg gui)"/>
  	<param name="config/concurrent_node_construction"  value="true"/>
  	<param name="config/concurrent_edge_construction"  value="true"/>
  	<param name="config/concurrent_optimization"       value="true"/>
    <param name="config/optimizer_skip_step"           value="$(arg optimizer_skip_step)"/><!--optimize at end only -->
    <param name="config/backend_solver"                value="cholmod"/>
    <param name="config/pose_relative_to"              value="inaffected"/>
    <param name="config/optimize_landmarks"            value="false"/> <!-- Include feature poses as vertices in graph optimization-->
  	<param name="config/data_skip_step"                value="1"/><!-- skip every n-th frame completely -->
  	<param name="config/visualization_skip_step"       value="1"/> <!-- draw only every nth pointcloud row and line, high values require higher squared_meshing_threshold -->
  	<param name="config/encoding_bgr"                  value="false"/>
  	<param name="config/visualize_mono_depth_overlay"  value="false"/>
    <param name="config/squared_meshing_threshold"     value="0.0081"/>
    <param name="config/batch_processing"              value="true"/>    <!--store results and close after bagfile has been processed-->
    <param name="config/keep_all_nodes"                value="true"/> <!-- add nodes with constant motion if no transformation can be found -->
    <param name="config/use_icp"                       value="false"/> <!-- Ignored if ICP is not compiled in (see top of CMakeLists.txt) -->
    <param name="config/gicp_max_cloud_size"           value="25000"/>
    <param name="config/use_root_sift"                 value="$(arg use_root_sift)"/>
    <param name="config/optimizer_iterations"          value="0.01"/><!-- maximum of iterations in online operation (i.e., does affect the final optimization in batch mode) -->
    <param name="config/emm__skip_step"                value="1"/> 
    <param name="config/emm__mark_outliers"            value="false"/> 
    <param name="config/observability_threshold"       value="$(arg observability_threshold)"/>  
  <!--      <param name="config/max_edge_error"       value="0.01"/>  -->
    <param name="config/cloud_creation_skip_step"      value="16"/> <!-- Only active if cloud is computed (i.e. "topic_points" is empty. This value multiplies to emm__skip_step and visualization_skip_step-->
    <param name="config/octomap_resolution"            value="0.01"/> <!-- Only active if cloud is computed (i.e. "topic_points" is empty. This value multiplies to emm__skip_step and visualization_skip_step-->
    <!--param name="config/g2o_transformation_refinement" value="1000"/--> <!-- Use g2o to refine the ransac result, i.e. optimize the Mahalanobis distance in a final step.-->
  </node>
</launch>
